src/Library/UEventDeviceManager.cpp

```diff
@@ -316,6 +316,90 @@ namespace usbguard
     }
   }
 
+  void UEventDeviceManager::scan(const std::string& devpath)
+  {
+    std::vector<std::string> components;
+    tokenizeString(devpath, components, "/", /*trim_empty=*/true);
+    auto base = std::find_if(components.begin(), components.end(),
+    [](const std::string& component) -> bool {
+      return hasPrefix(component, "usb");
+    });
+    std::string path = "";
+
+    for (auto itr = components.begin(); itr < components.end(); ++itr) {
+      path += "/" + *itr;
+
+      if (itr >= base) {
+        try {
+          std::unique_lock<std::mutex> lock(_enumeration_mutex);
+          uint32_t id = 0;
+          const bool known_path = knownSysfsPath(path, &id);
+          lock.unlock();
+
+          if (!known_path || id == 0) {
+            SysFSDevice sysfs_device(path);
+
+            /*
+             * Do some additional sanity checking.
+             */
+            if (sysfs_device.getUEvent().hasAttribute("DEVTYPE")) {
+              const std::string devtype = sysfs_device.getUEvent().getAttribute("DEVTYPE");
+
+              if (devtype != "usb_device") {
+                USBGUARD_LOG(Warning) << path << ": UEvent DEVTYPE != usb_device. Ignoring event.";
+                continue;
+              }
+            }
+            else {
+              if (!sysfs_device.hasAttribute("descriptors")) {
+                USBGUARD_LOG(Warning) << path << ": UEvent doesn't refer to a device with a descriptors file. Ignoring event.";
+                continue;
+              }
+            }
+
+            processDeviceInsertion(sysfs_device, /*signal_present=*/known_path);
+            USBGUARD_LOG(Debug) << "Enumeration notify: sysfs_devpath=" << path
+              << " _enumeration=" << _enumeration
+              << " known_path=" << known_path;
+          }
+        }
+        catch (const Exception& ex) {
+          USBGUARD_LOG(Warning) << "USB Device Exception: "
+            << ex.message();
+          DeviceException(ex.message());
+          return;
+        }
+        catch (...) {
+          USBGUARD_LOG(Warning) << "USB Device Exception: unknown exception";
+          DeviceException("unknown exception");
+          return;
+        }
+      }
+    }
+
+    try {
+      std::unique_lock<std::mutex> lock(_enumeration_mutex);
+      uint32_t id = 0;
+      const bool known_path = knownSysfsPath(devpath, &id);
+      lock.unlock();
+
+      if (known_path && id > 0) {
+        processDevicePresence(id);
+      }
+    }
+    catch (const Exception& ex) {
+      USBGUARD_LOG(Warning) << "USB Device Exception: "
+        << ex.message();
+      DeviceException(ex.message());
+      return;
+    }
+    catch (...) {
+      USBGUARD_LOG(Warning) << "USB Device Exception: unknown exception";
+      DeviceException("unknown exception");
+      return;
+    }
+  }
+
```

Comment:
This is a big code duplication. Would it be possible to refactor this and implement it somehow based on the existing scan method -- i.e. do a bit of refactoring to be able to share the common code?

Reply:
I was able to split ueventProcessUEvent in a way that allowed me to remove a big chunck of the duplicate code while not triggering uevents on the netlink socket.